File: computer-service/src/main/java/com/ddoscollector/computer/service/ComputerService.java
package com.ddoscollector.computer.service;

import com.ddoscollector.computer.model.Computer;

import java.util.List;
import java.util.Optional;

/**
 * This interface describes a computer service.
 */
public interface ComputerService {

    /**
     * Create a computer.
     *
     * @param computer the computer to create
     * @return the created computer
     */
    Computer createComputer(Computer computer);

    /**
     * Get computer.
     *
     * @param id the ID of the computer to get
     * @return an optional with computer if it exists or an empty optional otherwise
     */
    Optional<Computer> getComputer(String id);

    /**
     * Update computer.
     *
     * @param id                 the ID of the computer to update
     * @param computerWithUpdate the computer with the update
     * @return the updated computer
     */
    Computer updateComputer(String id, Computer computerWithUpdate);

    /**
     * Delete computer.
     *
     * @param id the ID of the computer to delete
     */
    void deleteComputer(String id);

    /**
     * Get computers.
     *
     * @return the computers
     */
    List<Computer> getComputers();
}

File: computer-ui/src/app/computer/service/computer.service.ts
import {Injectable} from "@angular/core";
import {Computer} from "../model/computer";
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs/internal/Observable";
import {SERVICE_URL} from "../../settings";

// todo put setting in config file/config service

@Injectable()
export class ComputerService {

  constructor(private http: HttpClient) {
  }

  saveComputer(computer: Computer): Observable<Computer> {
    if (computer.id == null) {
      return this.createComputer(computer);
    }
    return this.updateComputer(computer);
  }

  getComputer(id: string): Observable<Computer> {
    return this.http.get<Computer>(SERVICE_URL + "/" + id);
  }

  deleteComputer(id: string): Observable<Computer> {
    return this.http.delete<Computer>(SERVICE_URL + "/" + id)
  }

  getComputers(): Observable<Computer[]> {
    return this.http.get<Computer[]>(SERVICE_URL)
  }

  private createComputer(computer: Computer): Observable<Computer> {
    return this.http.post<Computer>(SERVICE_URL, computer)
  }

  private updateComputer(computer: Computer): Observable<Computer> {
    return this.http.patch<Computer>(SERVICE_URL + "/" + computer.id, computer)
  }
}
